tests/utilities/test_attributes.py

```diff
@@ -0,0 +1,90 @@
+        ((("1.0.2.3.0.4", 1), ("1.0.2", 0)), {"1": [{"2": 0, "2.3": [{"4": 1}]}]}),
+        ((("1.0.2.3.4.5", 1), ("1.0.2.3", 0)), {"1": [{"2": {"3": 0, "3.4": {"5": 1}}}]}),
+        ((("1.0.2.0.3.4", 1), ("1.0.2.0.3", 0)), {"1": [{"2": [{"3": 0, "3.4": 1}]}]}),
+        ((("1.2.0.3.0.4.5", 1), ("1", 0)), {"1": 0, "1.2": [{"3": [{"4": {"5": 1}}]}]}),
+        ((("1.2.3.0.4.0.5", 1), ("1.2", 0)), {"1": {"2": 0, "2.3": [{"4": [{"5": 1}]}]}}),
+        ((("1.0.2.3.0.4.5", 1), ("1.0.2", 0)), {"1": [{"2": 0, "2.3": [{"4": {"5": 1}}]}]}),
+        ((("1.0.2.3.4.0.5", 1), ("1.0.2.3", 0)), {"1": [{"2": {"3": 0, "3.4": [{"5": 1}]}}]}),
+        ((("1.0.2.0.3.4.5", 1), ("1.0.2.0.3", 0)), {"1": [{"2": [{"3": 0, "3.4": {"5": 1}}]}]}),
+        ((("1.0.2.0.3.4.5", 1), ("1.0.2.0.3.4", 0)), {"1": [{"2": [{"3": {"4": 0, "4.5": 1}}]}]}),
+        (
+            (("1.0.2.3.4.5.6", 2), ("1.0.2.3.4", 1), ("1.0.2", 0)),
+            {"1": [{"2": 0, "2.3": {"4": 1, "4.5": {"6": 2}}}]},
+        ),
+        (
+            (("0.0.0.0.0", 4), ("0.0.0.0", 3), ("0.0.0", 2), ("0.0", 1), ("0", 0)),
+            {"0": 0, "0.0": 1, "0.0.0": 2, "0.0.0.0": 3, "0.0.0.0.0": 4},
+        ),
+        (
+            (("a.9999999.c", 2), ("a.9999999.b", 1), ("a.99999.b", 0)),
+            {"a": [{"b": 0}, {"b": 1, "c": 2}]},
+        ),
+        ((("a", 0), ("c", 2), ("b", 1), ("d", 3)), {"a": 0, "b": 1, "c": 2, "d": 3}),
+        (
+            (("a.b.c", 0), ("a.e", 2), ("a.b.d", 1), ("f", 3)),
+            {"a": {"b": {"c": 0, "d": 1}, "e": 2}, "f": 3},
+        ),
+        (
+            (("a.1.d", 3), ("a.0.d", 2), ("a.0.c", 1), ("a.b", 0)),
+            {"a.b": 0, "a": [{"c": 1, "d": 2}, {"d": 3}]},
+        ),
+        (
+            (("a.0.d", 3), ("a.0.c", 2), ("a.b", 1), ("a", 0)),
+            {"a": 0, "a.b": 1, "a.0": {"c": 2, "d": 3}},
+        ),
+        (
+            (("a.0.1.d", 3), ("a.0.0.c", 2), ("a", 1), ("a.b", 0)),
+            {"a.b": 0, "a": 1, "a.0": [{"c": 2}, {"d": 3}]},
+        ),
+        (
+            (("a.1.0.e", 3), ("a.0.0.d", 2), ("a.0.0.c", 1), ("a.b", 0)),
+            {"a.b": 0, "a": [{"0": {"c": 1, "d": 2}}, {"0": {"e": 3}}]},
+        ),
+        (
+            (("a.b.1.e.0.f", 2), ("a.b.0.c", 0), ("a.b.0.d.e.0.f", 1)),
+            {"a": {"b": [{"c": 0, "d": {"e": [{"f": 1}]}}, {"e": [{"f": 2}]}]}},
+        ),
+    ],
+)
+def test_unflatten(key_value_pairs, desired):
+    actual = dict(unflatten(key_value_pairs))
+    assert actual == desired
+    actual = dict(unflatten(reversed(key_value_pairs)))
+    assert actual == desired
+
+
+@pytest.mark.parametrize("key_value_pairs,desired", [((("1.0.2", 0),), {"1": [{"2": 0}]})])
+def test_unflatten_separator(key_value_pairs, desired):
+    separator = str(random())
+    key_value_pairs = ((key.replace(".", separator), value) for key, value in key_value_pairs)
+    actual = dict(unflatten(key_value_pairs, separator=separator))
+    assert actual == desired
```

Comment:
For this test, can we avoid parameterization if we only have one pair? I do think it's preferable to avoid randomness in unit tests whenever possible (good discussion [here](https://softwareengineering.stackexchange.com/a/429603)). E.g., `separator = "-"` is better IMO.

Reply:
yea, that makes sense to me